src/v/cloud_storage/recovery_utils.cc

```diff
@@ -0,0 +1,143 @@
+ */
+
+#include "cloud_storage/recovery_utils.h"
+
+#include "cloud_storage/logger.h"
+
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+namespace {
+// The top level path for all recovery state files on the cloud storage bucket
+constexpr std::string_view recovery_result_prefix{"recovery_state"};
+
+// The format for recovery state files, containing the NTP, a random UUID and
+// the boolean status
+// {recovery_result_prefix}/{ns}/{topic}/{partition}_{uuid}/{success:true|false}
+constexpr std::string_view recovery_result_format{"{}/{}/{}/{}_{}.{}"};
+} // namespace
+
+namespace cloud_storage {
+
+ss::sstring
+generate_result_path(const storage::ntp_config& ntp_cfg, bool result) {
+    auto uuid = boost::uuids::random_generator()();
+    return fmt::format(
+      recovery_result_format,
+      recovery_result_prefix,
+      ntp_cfg.ntp().ns(),
+      ntp_cfg.ntp().tp.topic(),
+      ntp_cfg.ntp().tp.partition,
+      boost::uuids::to_string(uuid),
+      result);
+}
+
+ss::future<> place_download_result(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  const storage::ntp_config& ntp_cfg,
+  bool result_completed,
+  retry_chain_node& parent) {
+    retry_chain_node fib{&parent};
+    auto result_path = generate_result_path(ntp_cfg, result_completed);
+    auto result = co_await remote.upload_object(
+      bucket, cloud_storage_clients::object_key{result_path}, "", fib);
+    if (result != upload_result::success) {
+        vlog(
+          cst_log.error,
+          "failed to upload result to {}, error: {}",
+          result_path,
+          result);
+    }
+    co_return;
+}
+
+ss::future<std::vector<recovery_result>> gather_recovery_results(
+  remote& remote,
+  cloud_storage_clients::bucket_name bucket,
+  retry_chain_node& parent) {
+    retry_chain_node fib{&parent};
+    std::vector<recovery_result> results{};
+    auto result = co_await remote.list_objects(
+      bucket, fib, cloud_storage_clients::object_key{recovery_result_prefix});
+    if (result.has_error()) {
+        vlog(
+          cst_log.error, "failed to list recovery results: {}", result.error());
+        co_return results;
+    }
+
+    // Matches the recovery_result_format. Example:
+    // "recovery_state/test_ns/test_topic/0_<UUID>.true" OR
+    // "recovery_state/test_ns/test_topic/0_<UUID>.false"
+    std::regex result_expr{fmt::format(
+      "{}/(.*?)/(.*?)/(\\d+)_(.*?).(true|false)", recovery_result_prefix)};
+
+    results.reserve(result.value().size());
+    for (const auto& item : result.value()) {
+        std::cmatch matches;
+        if (std::regex_match(
+              item.key.begin(), item.key.end(), matches, result_expr)) {
+            results.push_back(
+              {model::topic_namespace{
+                 model::ns{matches[1].str()}, model::topic{matches[2].str()}},
+               model::partition_id{std::stoi(matches[3].str())},
```

Comment:
If it throws then it will be a problem, but I assume since we do a regex match on `\d` it would be safe and run `stoi` on the item which corresponds to the digit capture. 

I can add exception handling there though.